core-backend/sandbox/src/env.rs

```diff
@@ -369,58 +424,107 @@ where
 
         let Self {
             mut instance,
-            mut runtime,
             entries,
             entry_point,
+            mut store,
+            memory,
         } = self;
 
         let stack_end = instance
-            .get_global_val(STACK_END_EXPORT_NAME)
+            .get_global_val(&store, STACK_END_EXPORT_NAME)
             .and_then(|global| global.as_i32())
             .map(|global| global as u32);
 
-        let gas = runtime.ext.define_current_counter();
+        let gas = store
+            .data_mut()
+            .as_mut()
+            .unwrap_or_else(|| unreachable!("State must be set in `WasmiEnvironment::new`"))
+            .ext
+            .define_current_counter();
 
         instance
-            .set_global_val(GLOBAL_NAME_GAS, Value::I64(gas as i64))
+            .set_global_val(&mut store, GLOBAL_NAME_GAS, Value::I64(gas as i64))
             .map_err(|_| System(WrongInjectedGas))?;
 
+        let mut globals_provider = GlobalsAccessProvider {
+            instance: instance.clone(),
+            store: None,
+        };
+        let globals_provider_dyn_ref = &mut globals_provider as &mut dyn GlobalsAccessor;
+
+        // Pointer to the globals access provider is valid until the end of `invoke` method.
+        // So, we can safely use it inside lazy-pages and be sure that it points to the valid object.
+        // We cannot guaranty that `store` (and so globals also) will be in a valid state,
+        // because executor mut-borrows `store` during execution. But if it's in a valid state
+        // each moment when protect memory signal can occur, than this trick is pretty safe.
+        let globals_access_ptr = &globals_provider_dyn_ref as *const _ as HostPointer;
```

Comment:
looks like after this gets merged it is possible to implement wasmi host executor